File: src/main/java/org/worldscanner/export/JsonExporter.java
package org.worldscanner.export;

import org.worldscanner.model.SearchResult;

import java.io.BufferedWriter;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.util.List;

/**
 * Экспорт результатов поиска в JSON формат.
 */
public final class JsonExporter {

    private JsonExporter() {
    }

    public static void export(List<SearchResult> results, String outputPath) {
        if (outputPath == null || outputPath.isBlank()) {
            throw new IllegalArgumentException("Output path is required for JSON export.");
        }

        File outputFile = new File(outputPath);
        try (BufferedWriter writer = new BufferedWriter(new FileWriter(outputFile))) {
            writer.write("[");
            writer.newLine();
            for (int i = 0; i < results.size(); i++) {
                SearchResult result = results.get(i);
                writer.write("  {");
                writer.newLine();
                writer.write("    \"type\": \"" + escape(result.getBlockId() != null ? "block" : result.getItemId() != null ? "item" : "entity") + "\",");
                writer.newLine();
                writer.write("    \"identifier\": \"" + escape(result.getBlockId() != null ? result.getBlockId() : result.getItemId() != null ? result.getItemId() : result.getEntityId()) + "\",");
                writer.newLine();
                writer.write("    \"dimension\": \"" + escape(result.getDimension().name()) + "\",");
                writer.newLine();
                writer.write("    \"regionX\": " + result.getRegionX() + ",");
                writer.newLine();
                writer.write("    \"regionZ\": " + result.getRegionZ() + ",");
                writer.newLine();
                writer.write("    \"chunkX\": " + result.getX() + ",");
                writer.newLine();
                writer.write("    \"chunkZ\": " + result.getZ() + ",");
                writer.newLine();
                writer.write("    \"blockX\": " + result.getX() + ",");
                writer.newLine();
                writer.write("    \"blockY\": " + result.getY() + ",");
                writer.newLine();
                writer.write("    \"blockZ\": " + result.getZ() + ",");
                writer.newLine();
                writer.write("    \"additionalInfo\": \"scanned from region file\"");
                writer.newLine();
                writer.write("  }");
                if (i < results.size() - 1) {
                    writer.write(",");
                }
                writer.newLine();
            }
            writer.write("]");
            writer.newLine();
            writer.flush();
            System.out.println("JSON export complete: " + outputFile.getAbsolutePath());
        } catch (IOException e) {
            throw new RuntimeException("Failed to export JSON file", e);
        }
    }

    private static String escape(String value) {
        if (value == null) {
            return "";
        }
        return value.replace("\\", "\\\\").replace("\"", "\\\"").replace("\n", "\\n").replace("\r", "\\r");
    }
}

File: cli/src/main/kotlin/org/worldscanner/cli/JsonExporter.kt
package org.worldscanner.cli

import org.worldscanner.core.model.SearchResult
import java.io.BufferedWriter
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.nio.file.Path

object JsonExporter {

    fun export(results: List<SearchResult>, path: Path) {
        Files.newBufferedWriter(path, StandardCharsets.UTF_8).use { writer ->
            writer.write("[\n")
            for ((index, result) in results.withIndex()) {
                writer.write("  ${resultToJson(result)}")
                if (index < results.lastIndex) writer.write(",")
                writer.write("\n")
            }
            writer.write("]\n")
        }
    }

    private fun resultToJson(result: SearchResult): String {
        val source = result.source
        return buildString {
            append("{")
            append("\"item\":").append(jsonString(result.item.normalizedId)).append(",")
            append("\"count\":").append(result.item.count).append(",")
            append("\"dimension\":").append(jsonString(result.dimension.name.lowercase())).append(",")
            append("\"region\":[").append(result.regionX).append(",").append(result.regionZ).append("],")
            append("\"chunk\":[").append(result.chunkX).append(",").append(result.chunkZ).append("],")
            append("\"position\":")
            when (source) {
                is org.worldscanner.core.model.ItemSource.BlockEntity -> {
                    append("[")
                    append(source.pos.x).append(",")
                    append(source.pos.y).append(",")
                    append(source.pos.z)
                    append("]")
                    append(",\"source\":\"block_entity\",\"block\":").append(jsonString(source.id))
                }
                is org.worldscanner.core.model.ItemSource.Entity -> {
                    append("[")
                    append(source.pos?.x ?: 0).append(",")
                    append(source.pos?.y ?: 0).append(",")
                    append(source.pos?.z ?: 0)
                    append("]")
                    append(",\"source\":\"entity\",\"entity\":").append(jsonString(source.id))
                }
            }
            append(",")
            append("\"containerPath\":[")
            for ((i, container) in result.containerPath.withIndex()) {
                append(jsonString(container))
                if (i < result.containerPath.lastIndex) append(",")
            }
            append("]")
            append("}")
        }
    }

    private fun jsonString(value: String): String {
        val sb = StringBuilder(value.length + 2)
        sb.append('"')
        for (ch in value) {
            when (ch) {
                '"' -> sb.append("\\\"")
                '\\' -> sb.append("\\\\")
                '\n' -> sb.append("\\n")
                '\r' -> sb.append("\\r")
                '\t' -> sb.append("\\t")
                '\b' -> sb.append("\\b")
                '\u000C' -> sb.append("\\f")
                else -> if (ch < ' ') sb.append("\\u").append(ch.code.toString(16).padStart(4, '0')) else sb.append(ch)
            }
        }
        sb.append('"')
        return sb.toString()
    }
}
